### ISSUE TITLE
Allow all external download URLs to be aliased

### ISSUE THREAD

**Issue by rarkins** [containerbase/base#1067]
Some users need to build their own containerbase-derived images in their product environments which do not have direct internet access. Instead, they need to set up generic proxies on Artifactory for each distinct external host which is approved, such as github.com, nodejs.org, etc.

Here's some examples used by containerbase:

Host | Tool(s)
-- | --
https://nodejs.org | node
https://api.adoptium.net | java
https://downloads.lightbend.com | scala
https://github.com | sbt, python, php,
https://dot.net | dotnet
https://cache.ruby-lang.org/ | ruby

We can assume for now that such users can already redirect common registries like apt, npm, pip, etc - the primary concern is on arbitrary URLs like the above.

The best way to test this would be to have a build environment with external URLs blocked by default and then try to build a "full" image with all tools and find the missing URLs one by one.

The next thing we'd want is a simple way of defining alternative URLs using env.

One possibility would be to have a syntax where the full from/to is specified in separate, related variables. e.g.

`URL_REPLACE_0_FROM=https://node.org`
`URL_REPLACE_0_TO=https://artifactory.company.com/something/nested`

Such a syntax is verbose but then at least should cover every case.

A higher level approach would be like `JAVA_BASE_URL=https://artifactory.company.com/something/java` but has the downsides:
- One tool might try multiple hosts, and
- Multiple tools might use the same host (e.g. github.com)

---

**Supported Tools**
- [x] `docker` #1189 
- [x] `dart` #1197

**Comment by viceice** [comment]
will do this after conversation to higher level, as it's much easier to implement in typescript.

**Comment by rarkins** [comment]
I need this within maximum 4 weeks :-/

I think longer term we should take all these URLs and proxy them ourselves through our own CDN/host so that the user only needs to alias one or a few URLs and not 10+

**Comment by viceice** [comment]
blocked by
- https://github.com/containerbase/base/issues/771

**Comment by viceice** [comment]
**working**
- `npm` can be configured via `npm_config_registry` env or `~/.npmrc`
- `pip? can be configured via `PIP_INDEX_URL` env or  `~/.config/pip/pip.conf`
- `gem` can be configured via `RUBYGEMS_HOST` env or  `~/.gemrc` [^1] [^2]

[^1]: https://docs.inedo.com/docs/proget-feeds-rubygem
[^2]: https://guides.rubygems.org/command-reference/#gem-environment

**Comment by viceice** [comment]
related:
- #7

**Comment by randygeyer-ws** [comment]
Is it sufficient to replace just base urls or will the entire url path need to be set, perhaps minus the artifact file name/ext?

An alternate/simple way to test this is to add /etc/hosts entries for each default hostname to point back to localhost, vs actually blocking the outbound traffic.

**Comment by rarkins** [comment]
> Is it sufficient to replace just base urls or will the entire url path need to be set, perhaps minus the artifact file name/ext?

By base URL I'm not meaning only the host, I mean it could be like https://host.com/some/path too. Is that what you're worried about?

**Comment by randygeyer-ws** [comment]
Yes.  Good deal.
__________________________
Randy Geyer | Principal Solutions Architect, Mend | +1 214 926-4907 |
***@***.***

**Comment by viceice** [comment]
The linked PR above will fix this issue for all tools beside the tools installed via `gem`, `npm` or `pip`. Those would need more effort to override the registry urls.

Would it be enough to document how to use the specific package manager environment variables?

Otherwise i would convert those tools and update the default registry urls on the fly, but that needs some more time to implement.

**Comment by rarkins** [comment]
I think those tools hopefully have their own way of using own registries. I was after examples of redirecting URLs like GitHub and downloads.apache.org

**Comment by viceice** [comment]
will add the docs and also missing `gem`, `npm` and `pip` support

**Comment by rarkins** [comment]
@viceice does this require any more code, or just docs?

**Comment by viceice** [comment]
needs more code too

**Comment by EXHades** [comment]
please support redirect/replace `https://dl.google.com/go`

in some company intranet，unable to access `dl.google.com` directly

**Comment by viceice** [comment]
> please support redirect/replace `https://dl.google.com/go`
> 
> in some company intranet，unable to access `dl.google.com` directly 
> 
> 

already supported

**Comment by viceice** [comment]
when this build is done, we should have all tools compatible with url replace

https://github.com/containerbase/base/actions/runs/5964382723

**Comment by viceice** [comment]
We should mention the abilllity to change the default corepack registry (can only be done at runtime)

- https://github.com/nodejs/corepack/tree/main#environment-variables

**Comment by pmaieref** [comment]
Hello,
as far as I can see this should already work for "node". After encountering issues and some debugging I noticed that there is still a hardcoded node.jsorg value here: 
https://github.com/containerbase/base/blob/b9ffa658baca6b27ac1d687919b10c08a55c0989/src/cli/tools/node/resolver.ts#L14

Access to the "https://nodejs.org/dist/index.json" can't be overwritten as far as I can see even at runtime when it is needed to figure out the correct node version if "no version" or "latest" is passed.

Is it planned to be fixed within this issue or would you like me to create a new issue?

**Comment by viceice** [comment]
@pmaieref Please open a new discussion. that should already support replacement, see
https://github.com/containerbase/base/blob/c2a72c60e04d08b2527f264c4ff8fd30dd4cea7a/src/cli/services/http.service.ts#L167

**Comment by gergan** [comment]
> > please support redirect/replace `https://dl.google.com/go`
> > in some company intranet，unable to access `dl.google.com` directly
> 
> already supported

As I understand it- the new tools(v2) are downloading Golang from https://github.com/containerbase/ - https://github.com/containerbase/base/blob/da910b72d2b4acccc69f885b9b541c35825c6ed3/src/usr/local/containerbase/tools/v2/golang.sh#L43 and if it has the chance to hit the get_from_url function, the typescript code will replace the urls, but file_exists https://github.com/containerbase/base/blob/da910b72d2b4acccc69f885b9b541c35825c6ed3/src/usr/local/containerbase/tools/v2/golang.sh#L46C21-L46C32 which comes before it uses curl directly https://github.com/containerbase/base/blob/da910b72d2b4acccc69f885b9b541c35825c6ed3/src/usr/local/containerbase/utils/cache.sh#L301 and will fail the download. Do I need to write a new bug report about this or it could be handled in this bug? And what is the best way to resolve it - for now I just replaced the line with get_from_url and the check with -f on the file
